File: Advanced Graph Problems/Tree_Coin_Collecting_I.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations,combinations
from io import BytesIO, IOBase
from decimal import Decimal,getcontext
 
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
 
# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    while b:
        a,b = b,a%b
    return a
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
MI = lambda x : x(map(int, sys.stdin.readline().split()))
SI = lambda : sys.stdin.readline().strip()
SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##
 
#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lp,selfop): SegmentTree, bootstrap: bootstrap, Trie: tries
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, Segment Tree(lp): SegmentOther
#Graph1(dnc,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, nummat: matrix, SuffixAutomaton: sautomaton
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
# input_file = open(r'input.txt', 'r');sys.stdin = input_file
 
def extras():
    getcontext().prec = 50
    sys.setrecursionlimit(10**6)
    sys.set_int_max_str_digits(10**5)
# extras()
 
def interactive():
    import builtins
    # print(globals())
    globals()['print'] = lambda *args, **kwargs: builtins.print(*args, flush=True, **kwargs)
# interactive()
 
def GI(n,m=None,sub=-1,dirs=False,weight=False):
    if m==None:
        m = n-1
    d = [[] for i in range(n)]
    if not weight:
        for i in range(m):
            u,v = LII_C(lambda x:int(x)+sub)
            d[u].append(v)
            if not dirs:
                d[v].append(u)
    else:
        for i in range(m):
            u,v,w = LII()
            d[u+sub].append((v+sub,w))
            if not dirs:
                d[v+sub].append((u+sub,w))
    return d
 
ordalp = lambda s : ord(s)-65 if s.isupper() else ord(s)-97
alp = lambda x : chr(97+x)
yes = lambda : print("Yes")
no = lambda : print("No")
yn = lambda flag : print("Yes" if flag else "No")
printf = lambda x : print(-1 if x==float('inf') else x)
lalp = 'abcdefghijklmnopqrstuvwxyz'
ualp = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
dirs = ((1,0),(0,1),(-1,0),(0,-1))
dirs8 = ((1,0),(0,1),(-1,0),(0,-1),(1,-1),(-1,1),(1,1),(-1,-1))
ldir = {'D':(1,0),'U':(-1,0),'R':(0,1),'L':(0,-1)}
 
INF = float("inf")
class Dinic:
    def __init__(self, n):
        self.lvl = [0] * n
        self.ptr = [0] * n
        self.q = [0] * n
        self.adj = [[] for _ in range(n)]
    def add_edge(self, a, b, c, rcap=0):
        self.adj[a].append([b, len(self.adj[b]), c, 0])
        self.adj[b].append([a, len(self.adj[a]) - 1, rcap, 0])
    def dfs(self, v, t, f):
        if v == t or not f:
            return f
        for i in range(self.ptr[v], len(self.adj[v])):
            e = self.adj[v][i]
            if self.lvl[e[0]] == self.lvl[v] + 1:
                p = self.dfs(e[0], t, min(f, e[2] - e[3]))
                if p:
                    self.adj[v][i][3] += p
                    self.adj[e[0]][e[1]][3] -= p
                    return p
            self.ptr[v] += 1
        return 0
    def calc(self, s, t):
        flow, self.q[0] = 0, s
        for l in range(31):  # l = 30 maybe faster for random data
            while True:
                self.lvl, self.ptr = [0] * len(self.q), [0] * len(self.q)
                qi, qe, self.lvl[s] = 0, 1, 1
                while qi < qe and not self.lvl[t]:
                    v = self.q[qi]
                    qi += 1
                    for e in self.adj[v]:
                        if not self.lvl[e[0]] and (e[2] - e[3]) >> (30 - l):
                            self.q[qe] = e[0]
                            qe += 1
                            self.lvl[e[0]] = self.lvl[v] + 1
                p = self.dfs(s, t, INF)
                while p:
                    flow += p
                    p = self.dfs(s, t, INF)
                if not self.lvl[t]:
                    break
        return flow
 
class binary_lift:
    def __init__(self, graph, f=max, root=0, flag=False,data=[]):
        n = len(graph)
        parent = [-1] * (n + 1)
        depth = self.depth = [-1] * n
        bfs = [root]
        depth[root] = 0;flag2 = True
        if not data:data = [0]*n;flag2 = False
        for node in bfs:
            if not flag2:
                for nei,w in graph[node]:
                    if depth[nei] == -1:
                        data[nei] = w
                        parent[nei] = node
                        depth[nei] = depth[node] + 1
                        bfs.append(nei)
            else:
                for nei in graph[node]:
                    if depth[nei] == -1:
                        parent[nei] = node
                        depth[nei] = depth[node] + 1
                        bfs.append(nei)
        parent = self.parent = [parent]
        self.f = f
        if flag:
            data = self.data = [data]
            for _ in range(max(depth).bit_length()):
                old_data = data[-1]
                old_parent = parent[-1]
                data.append([f(val, old_data[p]) for val,p in zip(old_data, old_parent)])
                parent.append([old_parent[p] for p in old_parent])
    def lca(self, a, b):
        depth = self.depth
        parent = self.parent
        if depth[a] < depth[b]:
            a,b = b,a
        d = depth[a] - depth[b]
        for i in range(d.bit_length()):
            if (d >> i) & 1:
                a = parent[i][a]
        for i in range(depth[a].bit_length())[::-1]:
            if parent[i][a] != parent[i][b]:
                a = parent[i][a]
                b = parent[i][b]
        if a != b:
            return parent[0][a]
        else:
            return a
    def distance(self, a, b):
        return self.depth[a] + self.depth[b] - 2 * self.depth[self.lca(a,b)]
    def kth_ancestor(self, a, k):
        parent = self.parent
        if self.depth[a] < k:
            return -1
        for i in range(k.bit_length()):
            if (k >> i) & 1:
                a = parent[i][a]
        return a
    def __call__(self, a, b, c=0):
        depth = self.depth
        parent = self.parent
        data = self.data
        f = self.f
        c = self.lca(a, b)
        val = data[0][c]
        for x,d in (a, depth[a] - depth[c]), (b, depth[b] - depth[c]):
            for i in range(d.bit_length()):
                if (d >> i) & 1:
                    val = f(val, data[i][x])
                    x = parent[i][x]
        return val
    
def solve():
    n,q = LII()
    L = LII()
    d = GI(n)
 
    data = [-1]*n
 
    bfs = []
    for i in range(len(L)):
        if L[i]:
            data[i] = 0
            bfs.append((i,0))
        
    for elem,dist in bfs:
        for j in d[elem]:
            if data[j]==-1:
                data[j] = dist+1
                bfs.append((j,dist+1))
 
    gr = d
    n = len(gr)
    visited = [False]*n
    depth = [0]*n
    stack = [0]
    parent = [0]*n
    while stack:
        start = stack.pop()
        if not visited[start]:
            visited[start] = True
            for child in gr[start]:
                if not visited[child]:
                    depth[child] = depth[start]+1
                    parent[child] = start
                    stack.append(child)
 
    mini = []
    for i in range(n):
        mini.append(min(data[i],data[parent[i]]))
 
    mini = [mini]
    parent = [parent]
    
    for i in range(20):
        parent.append(parent[-1][:])
        mini.append(mini[-1][:])
        for j in range(n):
            parent[-1][j] = parent[-2][parent[-2][j]]
            mini[-1][j] = min(mini[-2][j],mini[-2][parent[-2][j]])
 
    def lca(a, b):
        if depth[a] < depth[b]:
            a, b = b, a
        d = depth[a] - depth[b]
        for i in range(20):
            if (d >> i) & 1:
                a = parent[i][a]
        if a == b:
            return a
        for i in reversed(range(20)):
            if parent[i][a] != parent[i][b]:
                a = parent[i][a]
                b = parent[i][b]
        return parent[0][a]
 
    def distf(a, b):
        c = lca(a, b)
        dist = depth[a] + depth[b] - 2 * depth[c]
        f = data[c]
        d = depth[a] - depth[c]
        x = a
        for i in range(20):
            if (d >> i) & 1:
                f = min(f, mini[i][x])
                x = parent[i][x]
        d = depth[b] - depth[c]
        x = b
        for i in range(20):
            if (d >> i) & 1:
                f = min(f, mini[i][x])
                x = parent[i][x]
        return dist + 2 * f
 
 
    for i in range(q):
        a,b = LII_1()
        # print(2*bl(a,b)+bl.distance(a,b))
        print(distf(a,b))
 
    #L1 = LII()
    #st = SI()
solve()

File: Advanced Graph Problems/Tree_Coin_Collecting_I.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;

#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vector<int>>;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")

template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<typename T>
class is_printable {
    template<typename U> static auto test(int) -> decltype(cout << declval<U>(), true_type{});
    template<typename> static auto test(...) -> false_type;
public: static constexpr bool value = decltype(test<T>(0))::value;
};
template<typename T> enable_if_t<is_printable<T>::value> __print(const T& val) { cout << val; }
template<typename T1, typename T2> void __print(const pair<T1, T2>& p) {
    cout << "("; __print(p.first); cout << ":"; __print(p.second); cout << ")";
}
template<typename T> enable_if_t<!is_printable<T>::value> __print(const T& container) {
    cout << "{"; bool first = true;
    for (const auto& x : container) { if (!first) cout << ", "; __print(x); first = false; }
    cout << "}";
}
void print() { cout << "\n"; }
template<typename T, typename... Args>
void print(const T& t, const Args&... rest) {
    __print(t); if constexpr (sizeof...(rest)) cout << " "; print(rest...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, SortedList: sortl
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

int lca(int a, int b, vector<int>& depth,vii& parent,vector<int>& data,vii& mini){
    if(depth[a]<depth[b]){
        int temp = a;
        a = b;
        b = temp;
    }
    int d = depth[a]-depth[b];
    for(int i=0;i<20;i++){
        if((d>>i)&1){
            a = parent[i][a];
        }
    }
    if(a==b){
        return a;
    }
    for(int i=19;i>=0;i--){
        if(parent[i][a]!=parent[i][b]){
            a = parent[i][a];
            b = parent[i][b];
        }
    }
    return parent[0][a];
}

int distf(int a, int b, vector<int>& depth,vii& parent,vector<int>& data,vii& mini){
    int c = lca(a,b,depth,parent,data,mini);
    int dist = depth[a]+depth[b]-2*depth[c];
    int f = data[c];
    int d = depth[a]-depth[c];
    int x = a;
    for(int i=0;i<20;i++){
        if((d>>i)&1){
            f = min(f,mini[i][x]);
            x = parent[i][x];
        }
    }
    d = depth[b]-depth[c];
    x = b;
    for(int i=0;i<20;i++){
        if((d>>i)&1){
            f = min(f,mini[i][x]);
            x = parent[i][x];
        }
    }
    return dist+2*f;
}

void solve() {
    int n,q;
    cin>>n>>q;

    vi L = LII(n);

    vector<vector<int>> d(n);

    for(int i=0;i<n-1;i++){
        int u,v;
        cin>>u>>v;
        u--;v--;
        d[u].pb(v);
        d[v].pb(u);
    }

    vector<int> data(len(L),-1);
    queue<pair<int,int>> bfs;
    for(int i=0;i<len(L);i++){
        if(L[i]){
            data[i] = 0;
            bfs.emplace(i,0);
        }
    }

    while(!bfs.empty()){
        auto &[elem,dist] = bfs.front();
        bfs.pop();
        for(auto j:d[elem]){
            if(data[j]==-1){
                data[j] = dist+1;
                bfs.emplace(j,dist+1);
            }
        }
    }

    n = len(d);
    vector<bool> visited(n,false);
    vi depth(n,0);

    stack<int> st;
    st.push(0);

    vector<int> parent1(n,0);
    while(!st.empty()){
        int start = st.top();
        st.pop();
        if(!visited[start]){
            visited[start] = true;
            for(auto child:d[start]){
                if(!visited[child]){
                    depth[child] = depth[start]+1;
                    parent1[child] = start;
                    st.push(child);
                }
            }
        }
    }

    vector<int> mini1;
    for(int i=0;i<n;i++){
        mini1.pb(min(data[i],data[parent1[i]]));
    }

    vector<vector<int>> mini;
    mini.push_back(mini1);

    vector<vector<int>> parent;
    parent.push_back(parent1);

    for(int i=0;i<20;i++){
        vi newparent(parent.back());
        vi newmini(mini.back());

        parent.pb(newparent);
        mini.pb(newmini);

        for(int j=0;j<n;j++){
            parent[len(parent)-1][j] = parent[len(parent)-2][parent[len(parent)-2][j]];
            mini[len(mini)-1][j] = min(mini[len(mini)-2][j],mini[len(mini)-2][parent[len(parent)-2][j]]);
        }
    }

    while(q--){
        int a,b;
        cin>>a>>b;
        a--;b--;
        cout<<distf(a,b,depth,parent,data,mini)<<endl;
    }

}

int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}